### PR TITLE
Add benchmarks

### PR DESCRIPTION
**Pull request by Vurich**
So I wrote these benchmarks with the aim of getting a vague idea of how Lightbeam compares to Cranelift in compile-time and runtime performance, you can see the results in this gist https://gist.github.com/Vurich/8696e67180aa3c93b4548fb1f298c29e.

I'd love any criticisms of the methodology used by these benchmarks w.r.t. the likely reliability of the results, I tried to dig into the code to make sure that, in the loop which is being measured, what I thought was being run was actually being run (and what I didn't think would be run was actually not being run), but I'd love if there were some holes to poke in the way that these benchmarks have been written.

Unfortunately, it changes the API of `wasmtime-wast` in order to make public a couple of functions (and one type returned by one of those functions) that the benchmarks need. I considered making these functions public only when `cfg(test)` is active, or to inline the benchmarks into that file itself instead of making a separate `benches` binary, but I figured that it would be more helpful to ask here what people would consider the best way to go about that. I personally didn't see anything majorly problematic with just making these methods public and it was the easiest method of implementation, so that's the method that I chose for the original version of this PR, with the understanding that I can change it later if necessary.

### CODE REVIEW

File: tests/all/wast.rs [hunk 1]
```diff
@@ -26,11 +38,12 @@ fn run_wast(wast: &str, strategy: Strategy) -> anyhow::Result<()> {
         .wasm_reference_types(reftypes)
         .wasm_multi_value(multi_val)
         .strategy(strategy)?
-        .cranelift_debug_verifier(true);
+        .cranelift_debug_verifier(cfg!(debug_assertions));
```
Comment (bjorn3) [inline tests/all/wast.rs:41]: Will this disable the verifier for fuzzing?

Comment (Vurich) [reply tests/all/wast.rs:41]: Great point, although I'm not sure if it actually will disable these checks in fuzzing - this file appears to only be for the spec test suite, and fuzzing should really have `debug_assertions` turned on anyway. Either way, I can easily revert this change, as I made it before I split the benchmarks into their own file - before, the tests and benchmarks were identical except that the benchmarks had `b.iter(|| {})` around `wast_context.run_file(wast)`. It's now not really necessary.

File: benches/main.rs [hunk 2]
```diff
@@ -0,0 +1,272 @@
+#![cfg(feature = "benches")]
+#![feature(test)]
```
Comment (alexcrichton) [inline benches/main.rs:2]: Could this file be renamed to perhaps `spec.rs` or something that indicates that it's benchmarking the spec test suite?

FWIW the spec test suite I don't think is really the most interesting of benchmarks, but at least for our own internal tracking and performance monitoring it probably isn't so bad to track!

File: benches/main.rs [hunk 3]
```diff
@@ -0,0 +1,272 @@
+#![cfg(feature = "benches")]
```
Comment (alexcrichton) [inline benches/main.rs:1]: Would it be possible to use a hand-rolled `harness = false` scheme or something like `criterion` that works on stable?

File: benches/main.rs [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+
+    let multi_val = path.iter().any(|s| s == "multi-value");
+
+    let mut cfg = Config::new();
+    cfg.wasm_simd(simd)
+        .wasm_bulk_memory(bulk_mem)
+        .wasm_reference_types(reftypes)
+        .wasm_multi_value(multi_val)
+        .strategy(strategy)?
+        .cranelift_debug_verifier(cfg!(debug_assertions));
```
Comment (alexcrichton) [inline benches/main.rs:56]: This I think is duplicated with the testing code as well as with the function below, could this all be consolidated into one function, perhaps in the `wasmtime-wast` crate itself?

File: benches/main.rs [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+    let adjust_wast = |mut err: wast::Error| {
+        err.set_path(path);
+        err.set_text(wast);
+        err
+    };
+
+    let file_contents = std::fs::read_to_string(path)
+        .with_context(|| format!("failed to read `{}`", path.display()))?;
+    let buf = ParseBuffer::new(&file_contents).map_err(adjust_wast)?;
+    let ast = parser::parse::<wast::Wast>(&buf).map_err(adjust_wast)?;
```
Comment (alexcrichton) [inline benches/main.rs:76]: Could this be deduplicated with the `wasmtime-wast` crate? Perhaps something where a function to handle each directive is provided and then we say "run this file", or something like that?

**Comment by alexcrichton** [comment]
Lightbeam was removed in https://github.com/bytecodealliance/wasmtime/pull/3390 as explained in [RFC 14](https://github.com/bytecodealliance/rfcs/pull/14), so I'm going to close this.

**Pull request closed without merging**